torch/csrc/jit/passes/integer_value_refinement.cpp

```diff
@@ -0,0 +1,199 @@
+#include <ATen/core/jit_type.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/integer_value_refinement.h>
+#include <torch/csrc/jit/passes/value_refinement_utils.h>
+#include <torch/csrc/utils/memory.h>
+#include "jit/ir/ir.h"
+
+namespace torch {
+namespace jit {
+
+using IntegerRefinement = std::unordered_map<Value*, int64_t>;
+
+// see [value refinement algorithm]
+
+struct IntegerValueRefiner {
+  IntegerValueRefiner(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  bool run() {
+    if (!checkForPossibleRefinements(graph_->block())) {
+      return false;
+    }
+    ListRefinement refinements;
+    RefineListLens(graph_->block(), refinements);
+    return changed_;
+  }
+
+  bool checkForPossibleRefinements(Block* b) {
+    for (Node* n : b->nodes()) {
+      if (n->matches("aten::eq(int a, int b) -> bool") ||
+          n->matches("aten::ne(int a, int b) -> bool")) {
+        for (size_t const_index : {0, 1}) {
+          auto non_const_index = 1 - const_index;
+          if (n->inputs().at(const_index)->node()->kind() == prim::Constant &&
+              n->inputs().at(non_const_index)->uses().size() > 1) {
+            return true;
+          }
+        }
+      }
+      for (Block* block : n->blocks()) {
+        if (checkForPossibleRefinements(block)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  ListRefinement RefineListLens(Block* b, ListRefinement block_refinements) {
+    active_refinements_.push_back(&block_refinements);
+    for (Node* n : b->nodes()) {
+      if (n->matches("aten::eq(int a, int b) -> bool") ||
+          n->matches("aten::ne(int a, int b) -> bool")) {
+        for (size_t const_index : {0, 1}) {
+          if (auto ival = constant_as<int64_t>(n->inputs().at(const_index))) {
+            IntegerRefinement refine;
+            refine[n->inputs().at(1 - const_index)] = *ival;
+            info_[n->output()] = n->kind() == aten::eq
+                ? BooleanRefinementMapping::TrueRefinements(std::move(refine))
+                : BooleanRefinementMapping::FalseRefinements(std::move(refine));
+          }
+        }
+      }
+      if (n->kind() == prim::RaiseException) {
+        throwing_blocks_.insert(b);
+      }
+
+      for (size_t input = 0; input < n->inputs().size(); ++input) {
```

Comment:
this does not apply to the other optimization, there we are optimizing `aten::len`, here we are optimizing all uses